pkg/security/resolvers/usersessions/resolver_linux.go

```diff
@@ -129,3 +170,363 @@ func (r *Resolver) ResolveUserSession(id uint64) *model.UserSessionContext {
+	// format for journalctl
+	sinceStr := ifr.lastRead.Format("2006-01-02 15:04:05")
+
+	cmd := exec.Command("journalctl", "--no-pager", "--since", sinceStr, "--output=short-iso")
+
+	var out bytes.Buffer
+	cmd.Stdout = &out
+
+	if err := cmd.Run(); err != nil {
+		seclog.Errorf("failed to read journalctl: %v", err)
+		return err
+	}
+
+	lines := strings.Split(strings.TrimSpace(out.String()), "\n")
+
+	var lastDate string
+	for i := 0; i < len(lines); i++ {
+		_, lastDate = parseSSHLogLine(lines[i], sshSessionParsed)
+	}
+	if lastDate == "" {
+		return nil
+	}
+	// We update the lastRead like this to avoid skipping another line that could be another ssh session
+	parsedSince, err := time.Parse("2006-01-02T15:04:05-0700", lastDate)
+	if err != nil {
+		seclog.Errorf("failed to parse date from journalctl: %v", err)
+	}
+	ifr.lastRead = parsedSince
+	return nil
+
+}
+
+// close closes the file.
+// The lock of IncrementalFileReader must be held
+func (ifr *incrementalFileReader) close() error {
+	if ifr.f != nil {
+		err := ifr.f.Close()
+		ifr.f = nil
+		return err
+	}
+	return nil
+}
+
+// inodeOf get the inode of the file.
+func inodeOf(fi os.FileInfo) uint64 {
+	if st, ok := fi.Sys().(*syscall.Stat_t); ok {
+		return st.Ino
+	}
+	return 0
+}
+
+// reloadIfRotated reopens the file if the inode has changed.
+func (ifr *incrementalFileReader) reloadIfRotated() error {
+	curSt, err := os.Stat(ifr.path)
+	if err != nil {
+		return err
+	}
+	curIno := inodeOf(curSt)
+	if curIno != 0 && ifr.ino != 0 && curIno != ifr.ino {
+		// The file has been rotated
+		if ifr.f != nil {
+			_ = ifr.close()
+			ifr.f = nil
+		}
+		f, err := os.Open(ifr.path)
+		if err != nil {
+			ifr.close()
+			ifr.f = nil
+			return err
+		}
+		ifr.f = f
+		ifr.ino = curIno
+
+		// We restart from the beginning because it's a new file
+		ifr.offset = 0
+	}
+	return nil
+}
+
+// StartSSHUserSessionResolver initializes the ssh log reader by looking for the available file, opening it and setting up the initial offset
+func (r *Resolver) StartSSHUserSessionResolver() {
+	possibleLogPaths := []string{
+		"/var/log/auth.log", // Debian/Ubuntu
+		"/var/log/secure",   // RHEL/CentOS/Fedora
+		"/var/log/messages", // openSUSE/autres
+	}
+	path := ""
+	var err error
+	for _, possiblePath := range possibleLogPaths {
+		_, err = os.Stat(possiblePath)
+		if err == nil {
+			path = possiblePath
+			break
+		}
+	}
+
+	r.sshLogReader = newIncrementalFileReader(path)
+
+	// Initialize the SSH session LRU cache (needed in all cases)
+	r.sshSessionParsed.lru, err = simplelru.NewLRU[sshSessionKey, sshSessionValue](100, nil)
+	if err != nil {
+		seclog.Errorf("couldn't create SSH Session LRU cache: %v", err)
+		return
+	}
+
+	if path == "" {
+		// Don't want to continue in case there is no log file, use journalctl instead
+		r.sshLogReader.lastRead = time.Now()
+		r.sshLogReader.readFromJournalctl = true
+		go r.startReading()
+		return
+	}
```

Comment:
This part mixes multiple things in ~wrong order, IMHO we should:
* start by initializing the LRU
* then do the loop
* make the `path == ""`
* an finally the `r.sshLogReader = newIncrementalFileReader(path)`